ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,334 @@
+                now,
+                NODE_QUERY_SATURATION,
+                saturation,
+                Collections.emptyMap()
+            )
+        )
+    );
+  }
+
+  private static final class KafkaStreamsSaturation {
+    private final QueryId queryId;
+    private final Map<String, ThreadSaturation> perThreadSaturation = new HashMap<>();
+    private final Duration window;
+    private final Duration sampleMargin;
+
+    private KafkaStreamsSaturation(
+        final QueryId queryId,
+        final Duration window,
+        final Duration sampleMargin
+    ) {
+      this.queryId = Objects.requireNonNull(queryId, "queryId");
+      this.window = Objects.requireNonNull(window, "window");
+      this.sampleMargin = Objects.requireNonNull(sampleMargin, "sampleMargin");
+    }
+
+    private void reportThreadSaturation(
+        final Instant now,
+        final double saturation,
+        final String name,
+        final MetricsReporter reporter
+    ) {
+      LOGGER.info("Reporting thread saturation {} for {}", saturation, name);
+      reporter.report(ImmutableList.of(
+          new DataPoint(
+              now,
+              QUERY_THREAD_SATURATION,
+              saturation,
+              ImmutableMap.of(THREAD_ID, name)
+          )
+      ));
+    }
+
+    private void reportQuerySaturation(
+        final Instant now,
+        final double saturation,
+        final QueryId queryId,
+        final MetricsReporter reporter
+    ) {
+      LOGGER.info("Reporting query saturation {} for {}", saturation, queryId);
+      reporter.report(ImmutableList.of(
+          new DataPoint(
+              now,
+              QUERY_SATURATION,
+              saturation,
+              ImmutableMap.of(QUERY_ID, queryId.toString())
+          )
+      ));
+    }
+
+    private Map<String, Map<String, Metric>> metricsByThread(final KafkaStreams kafkaStreams) {
+      return kafkaStreams.metrics().values().stream()
+          .filter(m -> m.metricName().group().equals(STREAMS_THREAD_METRICS_GROUP))
+          .collect(Collectors.groupingBy(
+              m -> m.metricName().tags().get(THREAD_ID),
+              Collectors.toMap(m -> m.metricName().name(), m -> m, (a, b) -> a)
+          ));
+    }
+
+    private Optional<Double> measure(
+        final Instant now,
+        final KafkaStreams kafkaStreams,
+        final MetricsReporter reporter
+    ) {
+      final Map<String, Map<String, Metric>> byThread = metricsByThread(kafkaStreams);
+      Optional<Double> saturation = Optional.of(0.0);
+      for (final Map.Entry<String, Map<String, Metric>> entry : byThread.entrySet()) {
+        final String threadName = entry.getKey();
+        final Map<String, Metric> metricsForThread = entry.getValue();
+        if (!metricsForThread.containsKey(STREAMS_TOTAL_BLOCKED_TIME)
+            || !metricsForThread.containsKey(STREAMS_THREAD_START_TIME)) {
+          LOGGER.info("Missing required metrics for thread: {}", threadName);
```

Comment:
If the thread is replaced after `metricsByThread` we would not see the metrics since they are GC'ed and would end here, is that right?

Reply:
We really should never hit this condition. If a thread is replaced then it won't be returned by `metricsByThread` at all.